### PR TITLE
enable FSDP example for model `hugging-quants/Meta-Llama-3.1-8B-Instr…

### PR DESCRIPTION
**Pull request by kaixuanliu**
Example cmd line:
`accelerate launch --config_file "fsdp_config.yaml"  train.py --seed 100 --model_name_or_path "hugging-quants/Meta-Llama-3.1-8B-Instruct-GPTQ-INT4" --dataset_name "smangrul/ultrachat-10k-chatml" --chat_template_format "chatml" --add_special_tokens False --append_concat_token False --splits "train,test" --max_seq_len 2048 --num_train_epochs 1 --logging_steps 5 --log_level "info" --logging_strategy "steps" --eval_strategy "epoch" --save_strategy "epoch" --bf16 True --packing True --learning_rate 1e-4 --lr_scheduler_type "cosine" --weight_decay 1e-4 --warmup_ratio 0.0 --max_grad_norm 1.0 --output_dir "llama-sft-lora-fsdp" --per_device_train_batch_size 1 --per_device_eval_batch_size 1 --gradient_accumulation_steps 4 --gradient_checkpointing True --use_reentrant False --dataset_text_field "content" --use_flash_attn False --use_peft_lora True --lora_r 8 --lora_alpha 16 --lora_dropout 0.1 --lora_target_modules "q_proj,k_proj,v_proj,o_proj,up_proj,gate_proj" --use_4bit_quantization False`

### CODE REVIEW

**Comment by kaixuanliu** [comment]
It depends on latest [GPTQModel](https://github.com/ModelCloud/GPTQModel) implementations. And 1 reminder: pls use transformers 4.52.4, the latest transformers has bug for this example, I am looking at this.

**Comment by kaixuanliu** [comment]
Hi, @BenjaminBossan , pls help review, thx!

**Comment by kaixuanliu** [comment]
Thx for advice, let's change back to original naming, as I noticed a lot of other places use `use_4bit_quantization`. It's better to keep the same.

File: examples/sft/utils.py [hunk 1]
```diff
@@ -178,7 +184,11 @@ def create_and_prepare_model(args, data_args, training_args):
         # ante). See https://github.com/huggingface/accelerate/issues/1620.
         uses_transformers_4_46 = packaging.version.parse(transformers.__version__) >= packaging.version.parse("4.46.0")
         uses_fsdp = os.environ.get("ACCELERATE_USE_FSDP", "false").lower() == "true"
-        if (bnb_config is not None) and uses_fsdp and uses_transformers_4_46:
+        if (
+            (bnb_config is not None or (hasattr(model, "hf_quantizer") and model.hf_quantizer is not None))
```
Comment (BenjaminBossan) [inline examples/sft/utils.py]: For better readability, let's assign this line to a variable like `is_quantized`, WDYT?

Comment (kaixuanliu) [reply examples/sft/utils.py]: Good advice! Have adjusted the code

**Comment by kaixuanliu** [comment]
I use Intel XPU to do the finetune. And here is my package version info:
```
torch                     2.9.0.dev20250629+xpu(nightly build)
transformers         4.52.4
trl                          0.19.0
accelerate             1.8.1
```

**Comment by kaixuanliu** [comment]
Along with accelerate config file I used:
```
compute_environment: LOCAL_MACHINE
debug: false
distributed_type: FSDP
downcast_bf16: 'no'
enable_cpu_affinity: false
fsdp_config:
  fsdp_activation_checkpointing: false
  fsdp_auto_wrap_policy: TRANSFORMER_BASED_WRAP
  fsdp_backward_prefetch: NO_PREFETCH
  fsdp_cpu_ram_efficient_loading: true
  fsdp_forward_prefetch: false
  fsdp_offload_params: false
  fsdp_reshard_after_forward: FULL_SHARD
  fsdp_state_dict_type: FULL_STATE_DICT
  fsdp_sync_module_states: true
  fsdp_transformer_layer_cls_to_wrap: ''
  fsdp_use_orig_params: false
  fsdp_version: 1
ipex_config:
  ipex: false
machine_rank: 0
main_training_function: main
mixed_precision: bf16
num_machines: 1
num_processes: 4
rdzv_backend: static
same_network: true
tpu_env: []
tpu_use_cluster: false
tpu_use_sudo: false
use_cpu: false
```

**Comment by kaixuanliu** [comment]
@BenjaminBossan ,Hi, I tested the case on A100, it can work as well. Here is my package version info:
```
torch                                         2.7.1+cu128
transformers                             4.52.4
trl                                              0.19.0
accelerate                                 1.8.1
CUDA                                        12.8
```
And cmd line:
`accelerate launch --config_file "configs/fsdp_config.yaml"  train.py --seed 100 --model_name_or_path "hugging-quants/Meta-Llama-3.1-8B-Instruct-GPTQ-INT4" --dataset_name "smangrul/ultrachat-10k-chatml" --chat_template_format "chatml" --add_special_tokens False --append_concat_token False --splits "train,test" --max_seq_len 2048 --num_train_epochs 1 --logging_steps 5 --log_level "info" --logging_strategy "steps" --eval_strategy "epoch" --save_strategy "epoch" --bf16 True --packing True --learning_rate 1e-4 --lr_scheduler_type "cosine" --weight_decay 1e-4 --warmup_ratio 0.0 --max_grad_norm 1.0 --output_dir "llama-sft-lora-fsdp" --per_device_train_batch_size 8 --per_device_eval_batch_size 8 --gradient_accumulation_steps 4 --gradient_checkpointing True --use_reentrant False --dataset_text_field "content" --use_flash_attn False --use_peft_lora True --lora_r 8 --lora_alpha 16 --lora_dropout 0.1 --lora_target_modules "q_proj,k_proj,v_proj,o_proj,up_proj,gate_proj" --use_4bit_quantization False`

File: examples/sft/utils.py [hunk 2]
```diff
@@ -178,7 +184,11 @@ def create_and_prepare_model(args, data_args, training_args):
         # ante). See https://github.com/huggingface/accelerate/issues/1620.
         uses_transformers_4_46 = packaging.version.parse(transformers.__version__) >= packaging.version.parse("4.46.0")
         uses_fsdp = os.environ.get("ACCELERATE_USE_FSDP", "false").lower() == "true"
-        if (bnb_config is not None) and uses_fsdp and uses_transformers_4_46:
+        is_quantized = bnb_config is not None
+        # Check if the model is already quantized by other approaches
+        if hasattr(model, "hf_quantizer") and model.hf_quantizer is not None:
+            is_quantized = True
```
Comment (BenjaminBossan) [inline examples/sft/utils.py]: How about this?

```suggestion
        # Check if the model is quantized
        is_quantized = (bnb_config is not None) or (getattr(model, "hf_quantizer", None) is not None)
```

**Comment by kaixuanliu** [comment]
Hi, I met the same issue with you and I have fixed it in https://github.com/ModelCloud/GPTQModel/pull/1642, pls use the latest GPTQModel(built from src code) by `pip install git+https://github.com/ModelCloud/GPTQModel.git` and it should can run successfully.

**Comment by BenjaminBossan** [comment]
> pls use the latest GPTQModel(built from src code) by `pip install git+https://github.com/ModelCloud/GPTQModel.git` and it should can run successfully.

I'm glad that I asked, this indeed resolved my problem.

Given this, I have a suggestion: How about extending the [README of this example](https://github.com/huggingface/peft/blob/main/examples/sft/README.md) with a new section on GPTQModel training. It should mention the min GPTQmodel version required (or installing from source) and include a working invocation (like you provided above). Possibly let's add a working `config.yaml` if the ones that are provided don't work. I think this will make it much more likely that users find this possibility and use it successfully. WDYT?

**Comment by kaixuanliu** [comment]
@BenjaminBossan Hi, I have added the README part, please help review. I double checked the existing `configs/fsdp_config.yaml` file can work in this case.

File: examples/sft/README.md [hunk 3]
```diff
@@ -27,7 +27,11 @@ When you have access to multiple GPUs, it would be better to use normal LoRA wit
 
 
 ## Multi-GPU SFT with LoRA and FSDP
-When you have access to multiple GPUs, it would be better to use normal LoRA with DeepSpeed/FSDP. To use LoRA with DeepSpeed, refer to the docs at [PEFT with FSDP](https://huggingface.co/docs/peft/accelerate/fsdp).
+When you have access to multiple GPUs, it would be better to use normal LoRA with DeepSpeed/FSDP. To use LoRA with FSDP, refer to the docs at [PEFT with FSDP](https://huggingface.co/docs/peft/accelerate/fsdp).
+
+
+## Multi-GPU SFT with LoRA and FSDP for GPTQModel:
+As in [Multi-GPU SFT with LoRA and FSDP](https://github.com/huggingface/peft/blob/main/examples/sft/README.md#multi-gpu-sft-with-lora-and-fsdp), we also support quantized model (Like GPTQModel) finetune. You may need to install [GPTQModel](https://github.com/ModelCloud/GPTQModel) later than v3.0.0 or from source code. Here is the launch command for reference: [run_peft_fsdp_gptq.sh]. And please note: there is a bug in transformers v4.53.0 for this case, please skip this transformers version.
```
Comment (BenjaminBossan) [inline examples/sft/README.md]:
```suggestion
As in [Multi-GPU SFT with LoRA and FSDP](https://github.com/huggingface/peft/blob/main/examples/sft/README.md#multi-gpu-sft-with-lora-and-fsdp), we also support other quantization methods like GPTQModel. You may need to install [GPTQModel](https://github.com/ModelCloud/GPTQModel) > v3.0.0 or from source. Here is the launch command for reference: [run_peft_fsdp_gptq.sh]. For the `--model_name_or_path` argument, it is important to pass a model that is already quantized with GPTQModel, like `"hugging-quants/Meta-Llama-3.1-8B-Instruct-GPTQ-INT4"`.

Note: there is a bug in transformers v4.53.0 for this case, please skip this transformers version.
```

As for the transformers bug, is it fixed in v4.53.1?

**Review by BenjaminBossan — COMMENTED** [review]
Thanks for the addition. I have a suggestion for a different wording in the README paragraph, WDYT?

**Comment by kaixuanliu** [comment]
@BenjaminBossan ,Hi using transformers with version 4.53.1 is OK.
